test/integration/authorize/list_app_keys_test.rb

```diff
@@ -0,0 +1,132 @@
+    assert !app_keys.nil?
+    assert_equal @application_nokeys.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert keys.empty?
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys even if no app_key is specified' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys even if app_key is invalid' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id,
+          app_key: 'invalid-key' },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys when specifying a valid app_key' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id,
+          app_key: @keys.sample },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
+  end
+
+  test "calling authorize with the list_app_keys extension lists a maximum of #{Transactor::Status.const_get(:LIST_APP_KEYS_MAX)} application keys" do
+    max_keys = Transactor::Status.const_get :LIST_APP_KEYS_MAX
+
+    (@keys.size + 1).upto(max_keys + 1) do |i|
+      @application.create_key "key_#{i}"
+    end
+
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id,
+          app_key: @keys.sample },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal max_keys, keys.size
+  end
+end
```

Comment:
It'd be good to add tests for services configured with user_key and oidc to verify that enabling the extension does not break anything and does not return anything.

Reply:
The idea is that the extension always returns something - at the very least the app and service the auth refers to, even if the app is user_key or oauth. We can test that at least once, but I suspect we already do somewhere else with an oauth application. Hmm.

Reply:
It does not make much sense to enable the extension for services configured with user_key, but I think that we should test that anyway and check that the list is empty and no exceptions are raised.

For OIDC is a bit more complicated. Services with OIDC should not have any app keys defined. However, because of what we discussed in a previous PR, it seems that Porta is storing them in apisonator. Should we return them? I don't think we should because it's not useful for caching anyway. The cache should only rely on the app ID for services that use OIDC. Whatever we decide, we should add a test for it.

Reply:
With user_key you can still make use of the extension to get the `app` and `svc` attributes of the response (`user_key` will map to an `app_id`), but admittedly a cache would just use directly the `user_key` value as a sort of `app_id` for caching.

For OIDC we are returning them and there is now a test for it.